src/Escher_Synth.ml

```diff
@@ -113,38 +114,54 @@ let solve_impl ?ast:(ast=false) task consts =
        List.iter ~f:(close_goal v) v_closes; true
   in
 
-  let int_components = List.filter ~f:(fun c -> c.codomain = TInt) components in
-  let bool_components = List.filter ~f:(fun c -> c.codomain = TBool) components in
-
   let apply_comp f types i =
     let rec apply_cells types acc locations = match types, locations with
-      | (typ::typs, i::locs) -> VSet.iter (fun x -> apply_cells typs (x::acc) locs) begin
-          match typ with
-            | TInt -> int_array.(i)
-            | TBool -> bool_array.(i)
-            | _ -> raise (Invalid_Type_Exn ("Escher does not handle type: "
+      | (typ::typs, i::locs) -> 
+      let apply_next = VSet.iter (fun x -> apply_cells typs (x::acc) locs) in begin
+        match typ with 
+          | TInt -> apply_next int_array.(i)
+          | TBool -> apply_next bool_array.(i)
+          | TList t -> apply_next list_array.(i)
+            (*
+            VSet.filter 
+            (fun (p, varray) -> (typeof varray.(0)) = (TList t)) list_array.(i) 
+            *)
+          | TVar s -> [int_array.(i); bool_array.(i); list_array.(i)] |> List.iter ~f:apply_next
+          | _ -> raise (Invalid_Type_Exn ("Escher does not handle type: "
                                            ^ (string_of_typ typ)))
-          end
+      end
       | ([], []) -> f (List.rev acc)
       | _ -> failwith "Impossible!"
     in divide (apply_cells types []) (List.length types) (i-1) []
   in
-  let expand_component c array i =
+  let expand_component c array_sel i =
     let f x =
+      (* We will modify c's codomain with the union result *)
+      try 
+      let typs = List.map ~f:(fun (_,x) -> typeof x.(0)) x in 
```

Comment:
Here as you can see, I get the type of the node by calculating the type of its first value on site. This can be inefficient, and i wonder if we can directly get the type of the values contained in the nodes?